R/rnet-clean.R

```diff
@@ -80,61 +84,174 @@
+  # A few safety checks
+  if (!inherits(rnet, "sf")) {
+    stop(
+      "At the moment this function was tested only using sf objects.",
+      call. = FALSE
+    )
+  }
+  if (!all(sf::st_is(rnet, "LINESTRING"))) {
+    stop(
+      "The input rnet must have LINESTRING geometry. See also ?sf::st_cast.",
+      call. = FALSE
     )
+  }
+  if (!is.null(sf::st_z_range(rnet)) || !is.null(sf::st_m_range(rnet))) {
+    warning(
+      "The Z/M dimensions will be lost.",
+      call. = FALSE,
+      immediate. = TRUE
+    )
+  }
 
-    message("Splitting rnet object at the shared boundary points.")
-    rnet_breakup_collection <- lwgeom::st_split(rnet, intersection_points$geometry)
-    rnet_clean <- sf::st_collection_extract(rnet_breakup_collection, "LINESTRING")
-  } else {
-    rnet_clean <- rnet
+  # Check if we need to rebuild tbl_df at the end
+  rebuild_tbl <- FALSE
+  if (inherits(rnet, "sf") && inherits(rnet, "tbl_df")) {
+    rebuild_tbl <- TRUE
   }
 
-  # Split again at the duplicated internal vertexes
-  rnet_internal_vertexes_duplicated <- rnet_internal_vertexes[duplicated(rnet_internal_vertexes)]
+  # step 1 - Find points that are both boundary points and internal points.
+  # 1a - Find internal points (i.e. points that are not in the boundary)
+  internal_points <- data.table::data.table(
+    sfheaders::sf_to_df(line2vertices(rnet))
+  )
+  data.table::setkey(internal_points, x, y)
+  # I don't need duplicated pairs of coordinates
+  internal_points <- unique(internal_points)
+  if (verbose) {
+    message("Extracted the (unique) internal points")
+  }
+  # 1b - Find points that lie in the boundaries
+  boundary_points <- data.table::data.table(
+    sfheaders::sf_to_df(rnet_boundary_unique(rnet))
+  )
+  data.table::setkey(boundary_points, x, y)
+  if (verbose) {
+    message("Extracted the (unique) boundary points")
+  }
 
-  if (length(rnet_internal_vertexes_duplicated) > 0 & breakup_internal_vertex_matches) {
-    message("Splitting rnet object at the shared internal points.")
-    rnet_breakup_collection <- lwgeom::st_split(rnet_clean, rnet_internal_vertexes_duplicated)
-    rnet_clean <- sf::st_collection_extract(rnet_breakup_collection, "LINESTRING")
+  # Consider only the points that are both internal and boundaries
+  shared_internal_points_data_table <- boundary_points[
+    # The following [ ] is used to filter the boundary points which are also
+    # internal points. The operation checks that the x and y coordinates
+    # are identical
+    stats::na.omit(boundary_points[internal_points, which = TRUE]),
+  ]
+
+  if (nrow(shared_internal_points_data_table) > 0) {
+    # step 2 - Split at shared internal points
+    message("Splitting the input object at the shared internal points.")
+    rnet <- my_st_split(rnet, shared_internal_points_data_table)
   }
 
-  rnet_clean
+  # step 3 - Find duplicated internal points
+  internal_points <- data.table::data.table(
+    sfheaders::sf_to_df(line2vertices(rnet))
```

Comment:
I don't think `line2vertices` is needed here - it's really slow. Replace with:

```r
     sfheaders::sf_to_df(rnet)
```

I suggest.

Reply:
I don't think that's true since `line2vertices()` is used to extract the "internal" points, i.e. those points that are part of the linestring(s) but do not lie in the boundaries. It's like the complementary of `rnet_boundary_points()`. I could rewrite also line2vertices using the sfheaders approach but, IMO, at the moment it's not worth it.

Reply:
Aha yes, I see you refactor that function: 

https://github.com/ropensci/stplanr/blob/master/R/od-funs.R#L456-L490

However... @agila5 you can get the internal and boundary points with a single operation as shown here, which I think is faster, so suggest calling `sfheaders::sf_to_df()`/` sf::st_coordinates()` (directly or indirectly) only once:


https://github.com/ropensci/stplanr/blob/8f055db02e07e83be18e81052ca971e5f9f394aa/data-raw/ad-hoc-tests/test-rnet-breakup-timing1.R#L33-L37

The `i` result here is the same as `line2vertices()` I think.

Reply:
> However... @agila5 you can get the internal and boundary points with a single operation as shown here, which I think is faster, so suggest calling sfheaders::sf_to_df()/ sf::st_coordinates() (directly or indirectly) only once

Yes, you are right, I will change that part!

Reply:
Should be fixed